src/bson_mapper/bson_archiver.hpp

```diff
@@ -0,0 +1,1013 @@
+#pragma once
+
+#include <memory>
+#include <stack>
+#include <string>
+#include <vector>
+
+#include <cereal/cereal.hpp>
+
+#include <bsoncxx/builder/core.hpp>
+#include <bsoncxx/types.hpp>
+#include <bsoncxx/types/value.hpp>
+
+namespace cereal {
+
+/**
+ * A base class that holds a shared_ptr to the binary data for a BSON document. If a class you are
+ * serializing contains any of the bsoncxx view types (b_utf8, b_document, b_array, b_binary), you
+ * must have that class inherit this base so that the views don't point to deallocated memory.
+ *
+ * When using the BSONInputArchive to deserialize a class that inherits from this base, the data
+ * will be automatically loaded into this class using setUnderlyingBSONData().
+ */
+class UnderlyingBSONDataBase {
+   public:
+    void setUnderlyingBSONData(std::shared_ptr<uint8_t> ptr, size_t size) {
+        _underlyingBSONData = ptr;
+        _underlyingBSONDataSize = size;
+    }
+
+    const bsoncxx::document::view getUnderlyingBSONData() {
+        return bsoncxx::document::view{_underlyingBSONData.get(), _underlyingBSONDataSize};
+    }
+
+   private:
+    std::shared_ptr<uint8_t> _underlyingBSONData;
+    size_t _underlyingBSONDataSize;
+};
+
+/**
+ * A templated struct containing a bool value that specifies whether the
+ * provided template parameter is a BSON type.
+ */
+template <class BsonT>
+struct is_bson {
+    static constexpr auto value = std::is_same<BsonT, bsoncxx::types::b_double>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_utf8>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_document>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_array>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_binary>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_oid>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_bool>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_date>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_int32>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_int64>::value ||
+                                  std::is_same<BsonT, bsoncxx::oid>::value;
+};
+
+/**
+ * A templated struct containing a bool value that specifies whether the
+ * provided template parameter is a BSON type that contains a view.
+ */
+template <class BsonT>
+struct is_bson_view {
+    static constexpr auto value = std::is_same<BsonT, bsoncxx::types::b_utf8>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_document>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_array>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_binary>::value;
```

Comment:
I specifically chose not to include those types as serializable with the ODM because in bsoncxx/types.hpp, these either have
```
/// @warning
///   This BSON type is used internally by the MongoDB server - use by clients
///   is discouraged.
```
or
```
/// @deprecated
///   A BSON DBPointer (aka DBRef) is still supported but deprecated.
```
I can add them if you think it would be a good idea, but I felt that we would be opening up an unnecessary can of worms if we started supporting these deprecated or internal types.